frontend/catalyst/pennylane_extensions.py

```diff
@@ -162,54 +181,108 @@ class QJITDevice(qml.QubitDevice):
+        return {
+            "Identity",
+            "PauliX",
+            "PauliY",
+            "PauliZ",
+            "Hadamard",
+            "S",
+            "T",
+            "PhaseShift",
+            "RX",
+            "RY",
+            "RZ",
+            "Rot",
+            "CNOT",
+            "CY",
+            "CZ",
+            "SWAP",
+            "IsingXX",
+            "IsingYY",
+            "IsingXY",
+            "ControlledPhaseShift",
+            "CRX",
+            "CRY",
+            "CRZ",
+            "CRot",
+            "CSWAP",
+            "Toffoli",
+            "MultiRZ",
+            "QubitUnitary",
+        }
+
+    @staticmethod
+    def _get_operations_to_convert_to_matrix(_config):
+        # We currently override and only set MultiControlledX to preserve current behaviour.
+        # We could choose to read from config and use the "matrix" gates.
+        # However, that affects differentiability.
+        # None of the "matrix" gates with more than 2 qubits parameters are differentiable.
+        # TODO: https://github.com/PennyLaneAI/catalyst/issues/398
+        return {"MultiControlledX"}
+
+    @staticmethod
+    def _check_mid_circuit_measurement(config):
+        return config["compilation"]["mid_circuit_measurement"]
+
+    @staticmethod
+    def _check_adjoint(config):
+        return config["compilation"]["adjoint"]
+
+    @staticmethod
+    def _check_quantum_control(config):
+        return config["compilation"]["quantum_control"]
+
+    @staticmethod
+    def _set_supported_operations(config):
+        """Override the set of supported operations."""
+        native_gates = set(config["operations"]["gates"][0]["native"])
+        qir_gates = QJITDevice._get_operations_available_in_QIR_abstract_machine()
+        QJITDevice.operations = list(native_gates.intersection(qir_gates))
+
+        # These are added unconditionally.
+        QJITDevice.operations += ["Cond", "WhileLoop", "ForLoop"]
+
+        if QJITDevice._check_mid_circuit_measurement(config):
+            QJITDevice.operations += ["MidCircuitMeasure"]
+
+        if QJITDevice._check_adjoint(config):
+            QJITDevice.operations += ["Adjoint"]
+
+        if QJITDevice._check_quantum_control(config):  # pragma: nocover
+            QJITDevice.operations += ["QCtrl"]  # pragma: nocover
+
+    @staticmethod
+    def _set_supported_observables(config):
+        """Override the set of supported observables."""
+        QJITDevice.observables = config["operations"]["observables"]
 
     # pylint: disable=too-many-arguments
     def __init__(
-        self, shots=None, wires=None, backend_name=None, backend_lib=None, backend_kwargs=None
+        self,
+        shots=None,
+        wires=None,
+        backend_name=None,
+        backend_lib=None,
+        backend_kwargs=None,
+        config=None,
     ):
+        QJITDevice._set_supported_operations(config)
+        QJITDevice._set_supported_observables(config)
```

Comment:
Looks like config is allowed to be `None`?

Reply:
Config is checked here https://github.com/PennyLaneAI/catalyst/pull/369/files/35662bd46c7ffb9eacd232d6a8515f7d4cfeecac#diff-4a57e2d034e69c6c5c19e4b6386ad5199dad536b909646a5906c2a2297f132d3R76-R90 and here https://github.com/PennyLaneAI/catalyst/pull/369/files/35662bd46c7ffb9eacd232d6a8515f7d4cfeecac#diff-4a57e2d034e69c6c5c19e4b6386ad5199dad536b909646a5906c2a2297f132d3R189-R193. I originally placed these functions inside QJITDevice but I didn't want QJITDevice to take another device as an input. All parameters to QJITDevice are obtained in the runtime.py file.

Reply:
But if `config` is a required parameter, why is it made optional in the signature?

Reply:
https://github.com/PennyLaneAI/catalyst/pull/369/commits/77778a83f6408c7ed09882e4a7a690e320df28d3